File: 8. String to Integer (atoi)/8. String to Integer (atoi).dart
class Solution {
  int myAtoi(String s) {
    int i = 0, sign = 1, result = 0;
    
    // Step 1: Ignore leading whitespace
    while (i < s.length && s[i] == ' ') i++;
    
    // Step 2: Check for sign
    if (i < s.length && (s[i] == '-' || s[i] == '+')) {
      sign = (s[i] == '-') ? -1 : 1;
      i++;
    }
    
    // Step 3: Convert digits
    while (i < s.length && s[i].codeUnitAt(0) >= '0'.codeUnitAt(0) && s[i].codeUnitAt(0) <= '9'.codeUnitAt(0)) {
      int digit = s[i].codeUnitAt(0) - '0'.codeUnitAt(0);
      // Check for overflow
      if (result > (2147483647 ~/ 10) || (result == (2147483647 ~/ 10) && digit > 2147483647 % 10)) {
        return (sign == 1) ? 2147483647 : -2147483648;
      }
      result = result * 10 + digit;
      i++;
    }
    
    return sign * result;
  }
}

File: 8. String to Integer (atoi)/8. String to Integer (atoi).py
class Solution(object):
    def myAtoi(self, s):
        """
        :type s: str
        :rtype: int
        """
        i, n = 0, len(s)
        
        # Step 1: Ignore leading whitespace
        while i < n and s[i] == ' ':
            i += 1
        
        # Step 2: Check for sign
        sign = 1
        if i < n and s[i] in ['-', '+']:
            sign = -1 if s[i] == '-' else 1
            i += 1
        
        # Step 3: Convert digits
        result = 0
        while i < n and s[i].isdigit():
            digit = int(s[i])
            # Check for overflow
            if result > 214748364 or (result == 214748364 and digit > 7):
                return 2147483647 if sign == 1 else -2147483648
            result = result * 10 + digit
            i += 1
        
        return sign * result